vendor/github.com/btcsuite/btcd/blockchain/blockindex.go

```diff
@@ -241,6 +242,7 @@ func newBlockIndex(db database.DB, chainParams *chaincfg.Params) *blockIndex {
 		db:          db,
 		chainParams: chainParams,
 		index:       make(map[chainhash.Hash]*blockNode),
+		dirty:       make(map[*blockNode]struct{}),
```

Comment:
This would make all of your function that touch `dirty` concurrently safe: https://godoc.org/golang.org/x/sync/syncmap